### PR TITLE
`Assessment`: Fix issues in the calculation of participant scores

### PR DESCRIPTION
**Pull request by ge65cer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

**Background:**
_Participant scores_ save the last (rated) _result_ for each combination of student/team and exercise.
Otherwise, we would have to go through the whole relation chain `Exercise -> Participation -> Submission -> Result` to find the last result for each student just to calculate the average course score for an exercise.

**Problem:**
Currently, we rely on callback methods from JPA entity lifecycle events (`@PostPersist`, `@PreRemove`, …). Then, we perform some database queries to fetch the participation and find the applicable last _result_.
However, Hibernate advises that [JPA callback methods](https://docs.jboss.org/hibernate/orm/5.0/userguide/en-US/html/ch12.html#d5e1611) must “not invoke EntityManager or Query methods!”. A recent issue #5633 demonstrates that this solution uses undefined behavior and may break easily. Also, every time a result is changed, the _participant score_ calculation runs and therefore slows down all requests modifying a result entity. Lastly, if the instance shuts down or an issue occurs during the calculation, the _participant score_ table does not repair itself and the value becomes invalid.

### Description

We move the calculation of _participant scores_ to a scheduler. The JPA callback methods send a message to the main instance to schedule the update of a _participant score_ for a modified _result_.
This ensures that the _participant scores_ are updated promptly within seconds.
We introduce a `last_modified_date` column for both _results_ and _participant scores_.
Using this date, a regular cron job schedules the processing of modified results. This also means that if the main instance goes down, the _participant scores_ can heal themselves the next time it boots up.

All in all, regardless of what happens, the _participant scores_ should eventually be consistent.

### Steps for Testing

Prerequisites:
- 1 Instructor, 1 Student
- 1 Course
- 1 Exercise of each type
- 1 Exam with an exercise of each type

1. Log in to Artemis
2. Navigate to Course Administration
3. Go to the Course and open the _"Participant Scores"_ in a new tab (**ensure it stays up-to-date for all tests below**)
4. Participate in each exercise as student
5. Assess the exercise as instructor 
6. Also try: Delete/Reset or edit the assessment, submission, participation or exercise
7. Participate in the exam as a student
8. Assess the exam as instructor 

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Instructions | Branches |
|------------|-------:|-----:|
| ResultListener.java | 94% | 75% |
| ParticipantScoreSchedulerService.java | 92% | 77% |

### CODE REVIEW

**Comment by b-fein** [comment]
Is this going to fix #5014, too?

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+        else {
+            participantScoreRepository.save(participantScore);
+            logger.debug("Updated participant score {}.", participantScore.getId());
+        }
+    }
+
+    /**
+     * Get the result that can replace the currently set last result for a participant score
+     *
+     * @author Stefan Waldhauser
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java:277]: Thats nice :D Thank you  ❤️

File: src/main/java/de/tum/in/www1/artemis/domain/LearningGoal.java [hunk 2]
```diff
@@ -26,7 +26,7 @@ public class LearningGoal extends DomainObject {
 
     @ManyToOne
     @JoinColumn(name = "course_id")
-    @JsonIgnoreProperties("learningGoals")
+    @JsonIgnoreProperties({ "learningGoals", "prerequisites" })
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/domain/LearningGoal.java:29]: ❓Seems unrelated to this PR?

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+            if (exercise.isTeamMode()) {
+                participant = teamRepository.findByIdElseThrow(participantId);
+                participantScore = teamScoreRepository.findByExercise_IdAndTeam_Id(exerciseId, participantId).map(score -> score);
+            }
+            else {
+                participant = userRepository.findByIdElseThrow(participantId);
+                participantScore = studentScoreRepository.findByExercise_IdAndUser_Id(exerciseId, participantId).map(score -> score);
+            }
+
+            if (participantScore.isPresent()) {
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java:208]: 💡 This method is rather long. Maybe you could extract some functionality into smaller methods? For example this part into a method like `checkIfUpdateIsNeeded` or something.

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+            if (result.getParticipation() instanceof StudentParticipation studentParticipation) {
+                var lastModified = result.getLastModifiedDate() == null ? Instant.now() : result.getLastModifiedDate();
+                scheduleTask(studentParticipation.getExercise().getId(), studentParticipation.getParticipant().getId(), lastModified);
+            }
+        });
+
+        var participantScoresToProcess = participantScoreRepository.findAllOutdated();
+        participantScoresToProcess.forEach(participantScore -> {
+            if (participantScore instanceof TeamScore teamScore) {
+                scheduleTask(teamScore.getExercise().getId(), teamScore.getTeam().getId(), Instant.now());
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java:133]: ❓If I understand the code correctly you are passing `Instant.now()` here because you want the below `already-up-to-date` check to fail. Just to confirm if I understood this part correctly

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]
Amazing job! Just a small problem with the liquibase config locally for me (see comment). Tested it with all course exercises except programming and it worked perfectly. Great Great jo!

File: src/main/java/de/tum/in/www1/artemis/domain/LearningGoal.java (hunk 2 above)
Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/domain/LearningGoal.java:29]: Needed to fix a failing test...

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java (hunk 4 above)
Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java:133]: Yes, exactly.

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java (hunk 3 above)
Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java:208]: It might be long, but it is not really complex (only one nested if-statement). If other reviewers agree, I'd like to keep it as-is.

File: src/main/resources/config/liquibase/changelog/20220913152104_changelog.xml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        <comment>Add a column to track the timestamp of the last update for results and participant scores.</comment>
+        <addColumn tableName="result">
+            <column name="last_modified_date" type="TIMESTAMP"/>
+        </addColumn>
+        <addColumn tableName="participant_score">
+            <column name="last_modified_date" type="TIMESTAMP"/>
+        </addColumn>
+
+        <createIndex tableName="result" indexName="result_last_modified_date">
+            <column>last_modified_date</column>
```
[the remark this answers is not in the document]
Comment (ge65cer) [reply src/main/resources/config/liquibase/changelog/20220913152104_changelog.xml:13]: Good catch! Fixed.

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java (hunk 3 above)
Comment (b-fein) [reply src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java:208]: +1 for splitting up `executeTask` and extracting parts into smaller helper methods from my side.

File: src/main/java/de/tum/in/www1/artemis/repository/StudentScoreRepository.java [hunk 6]
```diff
@@ -23,8 +23,8 @@ public interface StudentScoreRepository extends JpaRepository<StudentScore, Long
 
     void deleteAllByUser(User user);
 
-    @EntityGraph(type = LOAD, attributePaths = { "user", "exercise", "lastResult", "lastRatedResult" })
-    Optional<StudentScore> findStudentScoreByExerciseAndUser(Exercise exercise, User user);
+    @EntityGraph(type = LOAD, attributePaths = { "user", "exercise" })
+    Optional<StudentScore> findByExercise_IdAndUser_Id(Long exerciseId, Long userId);
```
Comment (simon-hng) [inline src/main/java/de/tum/in/www1/artemis/repository/StudentScoreRepository.java:27]: What's the rationale between this naming?

File: src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java [hunk 7]
```diff
@@ -648,18 +648,16 @@ public void deleteResultsAndSubmissionsOfParticipation(Long participationId) {
         var participation = participationRepository.findByIdWithResultsAndSubmissionsResults(participationId)
                 .orElseThrow(() -> new EntityNotFoundException("Participation", participationId));
         Set<Submission> submissions = participation.getSubmissions();
-        List<Result> resultsToBeDeleted = new ArrayList<>();
-
+        ArrayList<Result> resultsToBeDeleted = submissions.stream().flatMap(submission -> submission.getResults().stream()).collect(Collectors.toCollection(ArrayList::new));
```
Comment (simon-hng) [inline src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java:651]: nice usage of streams!

File: src/main/java/de/tum/in/www1/artemis/service/messaging/MainInstanceMessageSendService.java [hunk 8]
```diff
@@ -22,98 +23,87 @@ public MainInstanceMessageSendService(InstanceMessageReceiveService instanceMess
 
     @Override
     public void sendProgrammingExerciseSchedule(Long exerciseId) {
-        // No need to go through the broker, pass it directly
```
Comment (simon-hng) [inline src/main/java/de/tum/in/www1/artemis/service/messaging/MainInstanceMessageSendService.java:25]: interesting that those comment happened 🤔

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java (hunk 3 above)
Comment (simon-hng) [reply src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java:208]: Although smaller methods are in general better, I don't think it adds much value here

File: src/main/java/de/tum/in/www1/artemis/repository/StudentScoreRepository.java (hunk 6 above)
Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/repository/StudentScoreRepository.java:27]: Those are Spring JPA Data methods, you can read more about them here: https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.query-methods.details

File: src/main/java/de/tum/in/www1/artemis/service/messaging/MainInstanceMessageSendService.java (hunk 8 above)
Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/service/messaging/MainInstanceMessageSendService.java:25]: Jeah, probably through copy & paste.

**Review by simon-hng — APPROVED** [review]
Thank you, very cool

**Review by stefanwaldhauser — APPROVED** [review]
Addressed everything and tested it in test session again!

**Review by minrows — APPROVED** [review]
Tested on ts1 during testing session with author, LGTM 👍

**Review by akesfeden — APPROVED** [review]
Tested during testing session with author on ts1, scores are updated when they should and overall works as described.
